### DISCUSSION TITLE
Connecting IPython console to notebook

### DISCUSSION THREAD

**Discussion by AJajodia** [posit-dev/positron#3766] [Q&A] [answered]
I saw people being able to assign variables and access them in the same kernel instance (like RStudio) but in Python. For whatever reason, it's not working for me. I assign variables in either the notebook or the console and visually they stay in their own dropdown variables window. I am on MacOS Ventura 13.0.
![image](https://github.com/posit-dev/positron/assets/53088267/a3e5e424-cbf3-4285-bbed-25be6a82c6de)
![image](https://github.com/posit-dev/positron/assets/53088267/5c8e938e-7136-4821-85bc-2969049fdd78)

**Comment by seeM** [comment, marked as answer]
Variables aren't currently shared between the console session and a notebook session.

The console is intended to be used with ordinary Python scripts. You can press Cmd+Enter (or Ctrl+Enter) to execute the statement under the cursor in the console and advance through a script. You can also add `# %%` markers to denote "cells" in your scripts, and advance through those with Shift+Enter.

**Reply by AJajodia** [reply]
Ah thank you so much! Where would I find documentation for the added script functionality?

**Reply by seeM** [reply]
Our docs are currently at https://github.com/posit-dev/positron/wiki. Unfortunately, this functionality isn't yet documented.

It's similar to VSCode's Jupyter code cells so you might find [their docs](https://code.visualstudio.com/docs/python/jupyter-support-py#_jupyter-code-cells) useful in the meantime.

**Reply by ntluong95** [reply]
Why not share variables between the console session and a notebook session? Or at least having an option to attach that?

**Reply by seeM** [reply]
Our thinking was that most users would be confused by the console and notebook UI sharing the same session in the backend. Particularly because the console usually represents a different workflow (separate console + script panes) versus a notebook (which kinda combines those console and script panes into cells).

Can I ask when you'd use the console if you already have a notebook? For instance, why not execute those same commands in a notebook cell (which you can delete if you no longer need it)?

That said, an IDE that supports the console+script approach _and_ notebooks is quite a new thing, so we're happy to make refinements as we go. Please feel free to post this as an enhancement request issue!

**Reply by seeM** [reply]
Addressing https://github.com/posit-dev/positron/issues/2452 might be a better solution than sharing the console/notebook session. At the very least, it should make the UX much less confusing.

**Reply by AJajodia** [reply]
> Our thinking was that most users would be confused by the console and notebook UI sharing the same session in the backend. 

I made this thread for the exact opposite perspective as this one. I'm mostly an RStudio user, so I'm used to making .Rmd files and tinkering with their variables, doing summary statistics, etc. through the console. When I started having to use Python for work, I sought a way to do the same things I do in RStudio in my Jupyter Notebooks. To be honest, it's more confusing having separate variables. If it's possible, maybe a way to incorporate both perspectives would be best, like if the console dropdown could switch between a shared kernel and an independent instance. I'll add it as a feature request if it isn't already there, because I'm sure that other people would like this feature.

**Reply by seeM** [reply]
I see your point. It is a lot like editing an Rmd file while tinkering in the console, just with an ipynb file instead. Will discuss with the team!

**Reply by ntluong95** [reply]
I don't wan to make the Console below becoming useless. It is quite tempting to do some quick tinkering, totally throw away in the console
![image](https://github.com/posit-dev/positron/assets/63569637/ad8abb4d-9ac2-42e3-94db-01fa5da944e5)

**Reply by seeM** [reply]
To close the loop, here's the enhancement request: https://github.com/posit-dev/positron/issues/3801.
